refactor(core): Adjust our approach for upstream Zephyr boards

petejohanson opened the pull request:
Move to using proper HWMv2 board extensions
https://docs.zephyrproject.org/4.1.0/hardware/porting/board_porting.html#board-extensions

for extending upstream Zephyr boards for use with ZMK. With this change, using upstream Zephyr board IDs directly, e.g. `seeeduino_xiao` will be stock versions as found upstream. To use a board variant that is tuned for ZMK use, use the `zmk` variant, e.g. `seeeduino_xiao//zmk` which is shorthand for `seeeduino_xiao/samd21g18a/zmk`.

I've updated a few docs as part of this to note the variant change:

* https://deploy-preview-3145--zmk.netlify.app/blog/2025/12/09/zephyr-4-1#zmk-board-variant
* https://deploy-preview-3145--zmk.netlify.app/docs/hardware

<!-- Note: ZMK is generally not accepting PRs for new keyboards. New generic controller PRs *may* still be accepted, please discuss on the Discord server first. -->

## PR check-list

- [ ] Branch has a [clean commit history](https://zmk.dev/docs/development/contributing/pull-requests#clean-commit-history)
- [ ] Additional tests are included, if changing behaviors/core code that is testable.
- [ ] Proper Copyright + License headers added to applicable files (Generally, we stick to "The ZMK Contributors" for copyrights to help avoid churn when files get edited)
- [ ] [Pre-commit](https://zmk.dev/docs/development/local-toolchain/pre-commit) used to check formatting of files, commit messages, etc.
- [ ] Includes any necessary [documentation changes](https://zmk.dev/docs/development/contributing/documentation).

petejohanson committed:
refactor(core): Adjust our approach for upstream Zephyr boards

Move to using proper HWMv2 board extensions
https://docs.zephyrproject.org/4.1.0/hardware/porting/board_porting.html#board-extensions

for extending upstream Zephyr boards for use with ZMK. With this change,
using upstream Zephyr board IDs directly, e.g. `seeeduino_xiao` will be
stock versions as found upstream. To use a board variant that is tuned
for ZMK use, use the `zmk` variant, e.g. `seeeduino_xiao//zmk` which is
shorthand for `seeeduino_xiao/samd21g18a/zmk`.

petejohanson commented:
> I 100% agree with the guiding principles.
> 
> > So... If we do this, we perpetuate the existing challenges here, where you can't flash basic Zephyr samples to our "keyboard boards" from ZMK, without hacking apart the .dts files temporarily. Not a huge deal, but if we're putting in this effort, it might be worth improving this. I find this useful for targetted "get this display working" or "test the RGB" or "verify this sensor".
> 
> I could certainly be convinced to have //zmk variants for all of our boards, so that we _could_ flash Zephyr samples to them. My concern here is mainly around out-of-tree boards. I think for boards which are also interconnects, we can generally be convinced to add them in-tree of ZMK, if Zephyr does not accept them already, so my expectation is that out-of-tree boards are mostly just keyboards. Considering how much chaos and inconsistency there already is around out-of-tree boards, I see some value in keeping the expected definition for such boards simpler. I could really go either way, though.

Having tried this just the past couple days while playing with my stm32c0pad prototype, I can definitely say that having a "Zephyr compatible" based board for it was really helpful, and that adding a `zmk` variant was not a significant additional effort.

Not that I think we should expect everyone to do this with every board definition, but it might be nice to document it as an *option* for folks to follow.

Not sure how best to communicate this to users though? For sure, listing the `//zmk` variant in the .zmk.yml files is a good first start, with integration with the ZMK CLI helping here.
> 
> > Define "tweaks" here.
> 
> Essentially what you listed, though we also usually disable the UART node. I see an extra benefit of the separation as making it much easier to upstream boards to Zephyr since the definition has already been written, if a board gains popularity/prominence enough that it would be accepted.

:+1: 

> 
> > If we do this, I think we may introduce confusing for users, surrounding when and when not to use a zmk variant specifier.
> 
> I actually see the opposite: End users would be taught "If you have a board+shield combo, your board always has a //zmk at the end", without needing to look things up. The finer details would be more for advanced users.

I think my only real concern here, is where we draw the line for "accepting boards that support some interconnect", if we want to have that really be consistent/enforced.... The [lemon controllers](https://ryanis.cool/cosmos/lemon/) ? Halcyon? How about the tenth random RP2040 board we theoretically support? I'd love to have some planning around what that really looks like.

I do think that your proposed "user expectation" *does* seem like it would be nice to get to. How do we do so?

nmunnich commented:
> Not that I think we should expect everyone to do this with every board definition, but it might be nice to document it as an option for folks to follow.
> 
> Not sure how best to communicate this to users though? For sure, listing the //zmk variant in the .zmk.yml files is a good first start, with integration with the ZMK CLI helping here.
>
> I do think that your proposed "user expectation" does seem like it would be nice to get to. How do we do so?

Proposed plan:

- All in-tree boards get a //zmk variant
- We introduce a Kconfig flag that gets set in the variants (`CONFIG_ZMK`?). During build time we have the ZMK app check for this flag and throw a warning in the build process, something like "You are building ZMK firmware with a board that is missing ZMK-specific tweaks. You may experience some unexpected behavior. Have you checked if there is a <your_board>//zmk variant of your board?"
- The blog post and docs are updated to have //zmk everywhere, and we add a "patch note" to the blog post
- Probably we should adjust our supported hardware script to list boards which don't have in-tree shields as well.
- We add a "New board" page under hardware integration, which essentially states "Go to Zephyr and follow their instructions *exactly* to create a Zephyr board. <before continuing, here are some helpful Zephyr sample apps that you can flash to your board to ensure things are working at this point> Then return here for instructions on making a //zmk variant of your board with ZMK-specific tweaks allowing for Studio etc. If your board is a keyboard, once you've made the //zmk variant, take guidance (but don't follow exactly) the steps in the new shield guide to add a kscan etc." 

> I think my only real concern here, is where we draw the line for "accepting boards that support some interconnect", if we want to have that really be consistent/enforced.... The [lemon controllers](https://ryanis.cool/cosmos/lemon/) ? Halcyon? How about the tenth random RP2040 board we theoretically support? I'd love to have some planning around what that really looks like.

Looking at all of the boards we have already, and thinking about the ones which we want to keep in the long term, here's a draft proposal:

We definitely accept:

- Any ZMK variants of upstream boards
- Boards with interconnects which are individually purchasable from at least one reputable store and have gone through laboratory testing to become FCC/CE certified (independently, putting a certified module on a board does not suffice). 
- Boards with interconnects that are purchasable from multiple (3+) reputable stores, e.g. nice!nano, proton-c, elite-pi

We most likely accept:
- Open source boards with interconnects which are purchasable from at least one reputable store which isn't connected to its design. This covers mikoto, svlinky.

We likely accept:

- Open source boards with interconnects which are purchasable from at least one reputable store, that users will likely be using for custom keyboards. This covers lemon, upcoming version of splinktegrated.

We probably accept:

- Boards with interconnects which are individually purchasable from at least one reputable store and could be used for custom keyboards. By SplitKB's description of their future plans, it likely would include wired-only Halcyon in the future, but does not yet.

We maybe accept, but probably reject:

- Boards with interconnects that are individually purchasable but unlikely to be used for new designs, e.g. model-H.
- Boards with interconnects that are likely to be used for handwired designs only, e.g. Stampy.

We reject:

- Boards which feature technologies that we don't support (yet), e.g. half-duplex (current splinktegrated), anything 8bit (obviously)
- Boards with interconnects which are untested
- Boards with interconnects with limited availability/user count, e.g. vikoto
- Boards with interconnects which are only purchasable with an accompanying shield (current Halcyon)
- Boards with interconnects that are recreations of existing (likely closed source) in-tree boards, such as 0xCB Gemini
- Boards with interconnects that would match the above point, but the designer (for whatever reason) decided to break direct pin compatibility or make minor other changes. An extra LED or more memory doesn't suffice to not fall under this category, but the changes mikoto made to allow power input outside of USB or the addition of VIK to svlinky exclude those two here. 
- Boards with interconnects that are unsuitable for keyboards, e.g. a keyboard that is a board but is VIK-capable technically has an interconnect, but we don't want that in-tree.

If a board falls into multiple categories, then I think the lower category applies. So we reject all of the variants of the elite-pi that stores have made and open sourced (helios, liatris, etc). 

For boards which were once purchasable and were added to ZMK's tree, but are now no longer purchasable, I think we need to keep them in-tree for quite a while, but we can drop them the next time we need to do a significant board overhaul if they haven't been sold for an estimated...3 years? 5? 10? Pillbug, bluemicro840, makerdiary m.2. I don't think we should take responsibility for writing or maintaining a module for boards/shields which we remove from our tree, we should just drop them (with appropriate forewarning). We always keep the nrfmicro for historical reasons, and we keep the puchiBLE in-tree until it is no longer being sold, then treat it like the other 3.

petejohanson commented:
> Proposed plan:
> 
>     * All in-tree boards get a //zmk variant

Including "complete keyboards", that only ever are used with ZMK? We refactor so they can be used with Zephyr directly with the "stock" variant?

caksoylar commented:
It seems that (some?) board extensions are currently broken on main:
```sh
> west build -b sparkfun_pro_micro_rp2040 -d build/rp2040-test -- -DSHIELD=naked60
> grep CONFIG_ZMK_USB build/rp2040-test/zephyr/.config
# CONFIG_ZMK_USB is not set
```
@carrefinho also reported that board extensions are not applied on all boards whose name starts with vendor name, like `adafruit_kb2040`, `boardsource_blok`.

Do we want to fix that or get this in quicker?

petejohanson committed:
refactor(boards): Move to ZMK specific variants for the nRFMicro board

For consistency, adjust the nRFMicro board definition to offer a
"vanilla" Zephyr board, and then ZMK variants, e.g.
`nrfmicro/nrf52840/zmk`.

petejohanson committed:
refactor(boards): Move to ZMK specific variant for the bluemicro840 board

For consistency, adjust the bluemicro840 board definition to offer a
"vanilla" Zephyr board, and then ZMK variant, e.g. `bluemicro840//zmk`.

petejohanson committed:
refactor(boards): Make tofu65 to ZMK variant by default

Make the standard Tofu65 board ID be `tofu65/rp2040/zmk` or
`tofu65//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move BDN9 to ZMK variant by default

Make the standard BDN9 board ID be `bdn9/stm32f072xb/zmk` or
`bdn9//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move Puchi BLE to ZMK variant by default

Make the standard Puchi BLE board ID be `puchi_ble/nrf52840/zmk` or
`puchi_ble//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move Adv360 Pro to ZMK variant by default

Make the standard Adv360 Pro board ID be `adv360pro_left/nrf52840/zmk` or
`adv360pro_left//zmk` by shorthand and for right as well.

petejohanson committed:
refactor(boards): Move nRF52840 M2 to ZMK variant by default

Make the standard nRF52840 M2 board ID be `nrf52840_m2/nrf52840/zmk` or
`nrf52840_m2//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move Pillbug to ZMK variant by default

Make the standard Pillbug board ID be `pillbug/nrf52840/zmk` or
`pillbug//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move s40nc to ZMK variant by default

Make the standard s40nc board ID be `s40nc/nrf52840/zmk` or
`s40nc//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move nice!60 to ZMK variant by default

Make the standard nice!60 board ID be `nice60/nrf52840/zmk` or
`nice60//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move planck to ZMK variant by default

Make the standard planck board ID be `planck/stm32f303xc/zmk` or
`planck//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move preonic to ZMK variant by default

Make the standard preonic board ID be `preonic/stm32f303xc/zmk` or
`preonic//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move ferris to ZMK variant by default

Make the standard ferris board ID be `ferris/stm32f072xb/zmk` or
`ferris//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move Proton-C to ZMK variant by default

Make the standard Proton-C board ID be `proton_c/stm32f303xc/zmk` or
`proton_c//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move Corneish Zen to ZMK variant by default

Make the standard Corneish Zen board ID be `corneish_zen_left/nrf52840/zmk` or
`corneish_zen_left//zmk` by shorthand and for right as well.

petejohanson committed:
refactor(boards): Move nice!nano to ZMK variant by default

Make the standard nice!nano board ID be `nice_nano/nrf52840/zmk` or
`nice_nano//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move mikoto to ZMK variant by default

Make the standard mikoto board ID be `mikoto/nrf52840/zmk` or
`mikoto//zmk` by shorthand.

petejohanson committed:
refactor(boards): Move Polarity Works boards to ZMK variants

Make the standard Polarity Works board IDs be `zmk` variants.

petejohanson committed:
doc: Update docs/blog post to reference ZMK variants

* Update Zephyr 4.1 blog post to mention ZMK variants
* Add note to hardware support page about variants

nmunnich commented on app/boards/raspberrypi/rpi_pico/rpi_pico_rp2040_zmk.dts:
```diff
@@ -1,3 +1,10 @@
+/*
+ * Copyright (c) 2025 The ZMK Contributors
```
2026?

nmunnich commented on app/boards/adafruit/kb2040/board.yml:
Probably would be nice just to avoid all the help requests we'll likely be getting.

nmunnich reviewed (commented):
Are we leaving the "base" versions of full keyboards for later?

There are some references to `<shield>/boards/nice_nano.overlay` in the docs which could use renaming.

My comments on the serial bits and copyright years will apply to lots of places, putting them everywhere would be unhelpful I think.

petejohanson commented on docs/blog/2025-12-09-zephyr-4-1.md:
```diff
@@ -429,5 +435,13 @@ Once a ZMK version based on Zephyr 4.1 is released, we'll be working towards upd
 - Removing ZMK's use of the deprecated, and now removed, `kscan` APIs in favor of the newer matrix input API. ZMK already supports the matrix input API, but has not yet converted our existing drivers to that API. Advanced users looking to test can try leveraging the upstream Zephyr drivers today to test that functionality, but that is _not_ officially supported and is likely to have bugs or untes ...
 - Move the new USB stack, that includes better High-Speed USB support.
 
+:::note
```
Fair. Moved.

nmunnich commented on app/core-coverage.yml:
Nevermind, going commit by commit failed me on this file 😅 All good!

petejohanson committed:
docs: Fix up shield board overlays for new board IDs

Adjust our documentation to properly use the correct qualified board
overlay file names that match our new board conventions.

petejohanson commented on app/boards/extensions/xiao_ble/xiao_ble.conf:
```diff
@@ -1,19 +1,29 @@
-CONFIG_CONSOLE=n
-CONFIG_SERIAL=n
-CONFIG_UART_CONSOLE=n
-CONFIG_UART_INTERRUPT_DRIVEN=n
```
We're not. We're now:

* Leaving the original `xiao_ble` board definition untouched/changed.
* Adding a new `xiao_ble/nrf52840/zmk` variant, and that doesn't inherit the `xiao_ble_defconfig`, but has it's own, authoritative `xiao_ble_nrf52840_defconfig` file, which doesn't inherit from anything, so it doesn't have the serial/uart settings enabled simply by not having them included in that file.

petejohanson commented on app/boards/nordic/nrf52840dongle/nrf52840dongle_nrf52840_zmk_defconfig:
I'm not sure what adjustments you would expect? See my other comment here, which also applies here: https://github.com/zmkfirmware/zmk/pull/3145#discussion_r2750447340 when it comes to this no longer "inheriting" the defconfig file for the bare/non-variant board definition

petejohanson commented:
> I was indeed talking about non-ZMK versions of keyboards which are boards.

Tbh. This is lower priority for me. If the vendor/designer of a given board that's in tree wants to do it, they can submit a PR modeled after this one.

> 
> Unfortunate that the warnings couldn't be squashed. Could we open an issue with Zephyr to resolve that? Obviously not worth waiting for a fix from them, but I wouldn't want the warnings long-term.

Yeah. Fixing this just requires me to tweak how the included boot mode retention modified the RAM, which will squash that error. Will look at putting that together.

nmunnich reviewed (approved):
Well, minus the warning everything else looks fine to me, so I'll drop an approve and leave it up to you if you want to sort the warning now or later.

peterjc commented:
I mentioned this on Discord, but repeating here as it is more public: The blog post https://zmk.dev/blog/2025/12/09/zephyr-4-1 currently covers updating board names like `nice_nano_v2` in v0.3 -> `nice_nano@2.0.0` (or similar), but not this new change -> `nice_nano@2.0.0//zmk` (or similar).

I'd suggest updating the December 2025 blog post, or a second post for the Feb 2026 breaking changes (here in #3145). A second post would be best if any further changes from v0.3 are expected soon?

PeterMortensen commented:
### Broken links

The `deploy-preview-3145--zmk.netlify.app` links are broken (404). Would it be something like [this](https://zmk.dev/blog/2025/12/09/zephyr-4-1#zmk-board-variant) and [this](https://zmk.dev/docs/hardware)?

caksoylar commented:
The deploy previews are deployed from the PR branch (3145 is the github PR number, in this case). Since the PR is merged, you can instead check the paths in the main docs like the ones you shared, yes.
